main.ts

```diff
@@ -10,52 +9,148 @@ const connection = new (window as any).MapdCon()
+  prototypeData.transform = async function(_, pulse) {
+    console.log(_);
+    console.log(pulse);
+
+    const result = await session.queryAsync(_.query);
+    console.log(result);
+
+    const out = pulse.fork(pulse.NO_FIELDS & pulse.NO_SOURCE);
 
-  const result = await session.queryAsync(
-    `select airtime as "value" from ${table} where airtime is not null limit 1000`
-  );
+    out.add = out.source = result;
+    return out;
+  };
 
-  console.log(result);
+  /**
+   * A transform that sets a signals.
+   */
+  function MapDSignal(params) {
+    vega.Transform.call(this, null, params);
+  }
+  MapDSignal.Definition = {
+    type: "MapD_Signal",
+    metadata: { changes: true },
+    params: [
+      { name: "query", type: "string", required: true },
+      { name: "signal", type: "string", required: true },
+      { name: "array", type: "boolean" }
+    ]
+  };
+  const prototypeSignals = vega.inherits(MapDSignal, vega.Transform);
+  prototypeSignals.transform = async function(_, pulse) {
+    console.log(_);
+    console.log(pulse);
 
-  const { view } = await embed("#view", spec);
+    let result = await session.queryAsync(_.query);
 
-  view.insert("source", result).run();
+    if (_.array) {
+      // convert query result to an array
+      result = result.map(d => {
+        const arr = [];
+        Object.keys(d).forEach(k => {
+          arr[+k] = d[k];
+        });
+        return arr;
+      });
+
+      // we treat results with one row as a flat table
+      if (result.length === 1) {
+        result = result[0];
+      }
+    }
+
+    console.log(result);
+    this.value = result;
+  };
+
+  // add mapd transforms
+  vega.transforms["mapd_data"] = MapDData;
+  vega.transforms["mapd_signal"] = MapDSignal;
+
+  const runtime = vega.parse(spec);
+  const view = new vega.View(runtime)
+    .initialize(document.querySelector("#view"))
+    .run();
+
+  console.log(view);
 }
 
-const spec: Spec = {
+// transform to compute the extent
+const extentMapD = {
+  type: "mapd_signal",
+  query: {
+    signal: `'select min(airtime) as "0", max(airtime) as "1" from ${table}'`
+  },
+  signal: "extent",
+  array: true
+} as any;
+
+// transform to compute the extent
+const dataMapD = {
+  type: "mapd_data",
+  query: {
+    signal: `'select airtime as "value" from ${table} where airtime is not null limit 1000'`
+  }
+} as any;
+
+const spec: vega.Spec = {
   $schema: "https://vega.github.io/schema/vega/v4.json",
   autosize: "pad",
   padding: 5,
   width: 600,
   height: 250,
+  signals: [
+    { name: "maxbins", value: 20, bind: { min: 1, max: 200, type: "range" } }
+  ],
   data: [
-    { name: "source" },
+    {
+      name: "source",
+      transform: [dataMapD]
+    },
+    {
+      name: "extent",
+      transform: [extentMapD]
+    },
     {
       name: "table",
       source: "source",
       transform: [
-        {
-          type: "extent",
-          field: "value",
-          signal: "extent"
-        },
+        // {
```

Comment:
Stick your extentMapD transform here instead? Make sure the transform sets `this.value` to the extent array, then the runtime will bind that value to the `extent` signal for you after the transform has completed.

Reply:
If you include your `extentMapD` transform in the midst of a transform pipeline, you will need it pass the input pulse through as the output. You could also put it in it's own data source, in which case it could act like `dataMapD`, returning the query results as ingested Vega tuples.

The key bit is to either (a) set `transform.value` to the desired extent array and include a `signal` property on the transform to have that value copied to a named signal or (b) if the extent data is its own data source, you can define a separate signal that accesses that data source using the `data` expression function and constructs the extent range directly (e.g., `"update": "[data('extent')[0]['min'], data('extent')[0]['max']]"`).

Note that if you follow option (b) you only need to register one new custom transform, responsible for data loading and nothing else.

Also note that your "trick" of using "0" and "1" as the property names to fake an array might fail downstream, as numerous array processing methods in Vega assume a "length" property is also available.

Reply:
> If you include your `extentMapD` transform in the midst of a transform pipeline, you will need it pass the input pulse through as the output.

If I do that, wouldn't I compute the extent every single time a pulse comes through my transform?

> (b) if the extent data is its own data source, you can define a separate signal that accesses that data source using the `data` expression function and constructs the extent range directly (e.g., `"update": "[data('extent')[0]['min'], data('extent')[0]['max']]"`).

I like this idea and will see how it affects the flow.

> Also note that your "trick" of using "0" and "1" as the property names to fake an array might fail downstream, as numerous array processing methods in Vega assume a "length" property is also available.

Note that I actually converted the object into an array. I'm getting rid of this hack, though.